src/merkle/tiered_smt/mod.rs

```diff
@@ -0,0 +1,211 @@
+use super::{BTreeMap, Felt, Rpo256, RpoDigest, Serializable, Vec, Word};
+
+mod bits;
+pub use bits::BitsIterator;
+
+mod content;
+pub use content::{Content, ContentType};
+
+mod index;
+pub use index::TreeIndex;
+
+mod storage;
+pub use storage::{Storage, StorageError};
+
+#[cfg(test)]
+mod tests;
+
+// TIERED SPARSE MERKLE TREE
+// ================================================================================================
+
+pub struct TieredSmt {
+    // TODO this should be a constant
+    empty_subtrees: Vec<RpoDigest>,
+    storage: Storage,
+}
+
+impl TieredSmt {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    pub const MAX_DEPTH: usize = 64;
+    pub const TIER_DEPTH: usize = 4;
+
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn with_storage(storage: Storage) -> Self {
+        // Construct empty node digests for each layer of the tree
+        let empty_subtrees = (0..Self::MAX_DEPTH + 1)
+            .scan(Word::default().into(), |state, _| {
+                let value = *state;
+                *state = Rpo256::merge(&[value, value]);
+                Some(value)
+            })
+            .collect::<Vec<_>>()
+            .into_iter()
+            .rev()
+            .collect();
+
+        Self {
+            empty_subtrees,
+            storage,
+        }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn peek_node_type(&self, index: &TreeIndex) -> Result<ContentType, StorageError> {
+        self.storage
+            .peek_node_type(index)
+            .map(Option::unwrap_or_default)
+    }
+
+    pub fn get_node_digest(&self, index: &TreeIndex) -> Result<RpoDigest, StorageError> {
+        self.storage
+            .get_node_digest(index)?
+            .or_else(|| self.empty_subtrees.get(index.depth() as usize).copied())
+            .ok_or(StorageError::InvalidTreeIndex)
+    }
+
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn insert<T>(&self, value: &T) -> Result<RpoDigest, StorageError>
```

Comment:
I am not sure this is correct. The SMT is key-value map where the key is independent of the value.

The key is a 4-element word, but we use only the first element for indexing. If the two keys collide on the first element, we just put the corresponding key-value pairs into a sorted list in a node at depth 64.

The value is also a 4-element word. There is one open question about whether we want to somehow have different type of values (i.e., a single word or a list of words) - but for now lets keep things simple.

Overall, I think the `insert` function could have the following signature:
```Rust
pub fn insert(&mut self, key: Word, value: Word) -> Result<(), StorageError>
```

Reply:
Understood! It might be worthy to add this as documentation.

Usually, it is desirable to have a function `F` that, given a sparse merkle tree `T`, and an abitrary value `X`, we check the membership of `X` on `T`. To achieve that, the key must be a counter-domain of `X` of a collision-resistant function (i.e. hash).

Provided we allow an arbitrary key `K` for `T`, we need to change `F(X, T)` to `F(X, K, T)`. What would be good is to put a succinct documentation note for `insert` explaining why we allow the user to provide arbitrary keys.

Reply:
Not sure we need any special docs here. If we do want to add docs describing this behavior we should do it at a higher level as both `MerkleTree` and `SimpleSmt` behave the same way.